src/provider/utils/async.ts

```diff
@@ -81,6 +91,39 @@ export function allowReject<T>(promise: Promise<T>): Promise<T> {
     return promise;
 }
 
+/**
+ * Takes multiple promises representing API calls, and reduces them to a single result. In the case that more than one promise resolves to
+ * a result, the first to result will be used. Intented to be used when calling a client from the provider, to protect against a
+ * misbehaving client.
+ *
+ * @param promises An array of promises
+ */
+export async function collateApiCallResults<T = void>(promises: Promise<T>[]): Promise<[CollateApiCallResultsResult, T | undefined]> {
```

Comment:
Is resolving in the case of an error resulting in a bit of an anti-pattern here?..

I understand that the calling code will need to add context to any failures that occur, but would that be better done by this throwing an `FDC3Error` that can be (optionally) caught and decorated further up the stack?

I guess it's largely a question of semantics, would need to see which approach looks better when actually used in practice(?)

Reply:
With this I'd rather force the caller to consider failure, which they'd be able to just ignore if this throws exceptions. And I don't think they'd be any case where would could just let the error bubble up